### ISSUE TITLE
Why autoencoder does reduce_mean without specifying axis?

### ISSUE THREAD

**Issue by gabrieledcjr** [ageron/handson-ml#356]
In the autoencoder code, the loss is computed by:

`reconstruction_loss = tf.reduce_mean(tf.square(outputs - X))`

Just wanted some clarification. tf.reduce_mean gets the mean pixel by pixel, not from each sample from each batch. Should it be 

`reconstruction_loss = tf.reduce_mean(tf.reduce_sum(tf.square(outputs - X), axis=1))`

or

`reconstruction_loss = tf.reduce_mean(tf.reduce_mean(tf.square(outputs - X), axis=1))`

This way, it gets the mean or sum of the squared error for each sample in the batch and then get the mean error of the batch. Current implementation seems to get the mean of the squared error of all pixels from all samples of the batch.

**Comment by ageron** [comment]
Hi @gabrieledcjr ,
Interesting question. It turns out that both solutions give the same result (try it out). Indeed, computing the mean over every cell in a matrix gives the same result as first computing the mean over each row and then computing the mean over those means.

For example:

```pycon
>>> import numpy as np
>>> a = np.array([[1, 6, 8], [6, 7, 8]])
>>> a.mean()   # overall
6.0
>>> a.mean(axis=1)  # mean per row
array([5., 7.])
>>> a.mean(axis=1).mean()   # mean of the means == overall mean
6.0
```

Hope this helps!

**Comment by ageron** [comment]
To understand why, consider an array `a` with `m` rows and `n` columns. The mean of the i-th row is equal to `(a[i, 0] + a[i, 1] + ... + a[i, n-1]) / n`. The mean of the means is therefore equal to `((a[0, 0] + a[0, 1] + ... + a[0, n-1]) / n + (a[1, 0] + a[1, 1] + ... + a[1, n-1]) / n + ... + (a[m-1, 0] + a[m-1, 1] + ... + a[m-1, n]) / n) / m`. We can factor out the 1/n terms, and this gives us: `(a[0, 0] + a[0, 1] + ... + a[0, n] + a[1, 0] + a[1, 1] + ... + a[1, n] + ... + a[m-1, 0] + a[m-1, 1] + ... + a[m-1, n-1]) / (m*n)`. As you can see, this is equal to the sum of all the cells in the matrix, divided by the number of cells in the matrix (`m*n`), which is precisely the definition of the overall mean of the matrix. In short, it does not matter whether you compute the mean in one shot, overall, or whether you first compute the mean of each row (or each column), and then compute the mean of the means.
